### ISSUE TITLE
[Bug]: TA locks up entire server and other continers during download. 

### ISSUE THREAD

**Issue by CorneliousJD** [tubearchivist/tubearchivist#402]
### I've read the documentation

- [X] I'm running the latest version of Tube Archivist and have read the [release notes](https://github.com/tubearchivist/tubearchivist/releases/latest).
- [X] I have read through the [wiki](https://github.com/tubearchivist/tubearchivist/wiki) and the [readme](https://github.com/tubearchivist/tubearchivist#installing-and-updating), particularly the [common errors](https://github.com/tubearchivist/tubearchivist#common-errors) section.

### Operating System

unRAID

### Your Bug Report

Hi all, I have been running TubeArchivist now for a few weeks without issue, just subscribed to one channel or and downloading one video here or there. I added a few channels recently with ~500 videos or so ending up in my download queue.

Now all of a sudden every time I start the download process after a while my entire unraid server starts to become unresponsive, which is VERY weird considering this is just a container... 

I have tried limiting the download speed to 5000kbps (I have 500Mbps download connection) but the same thing still happens.
Same thing with pinning the container to only use 2 cores, but same thing occurs.

It's so bad that I can't even stop the container, unraid just says "server error" when I try and I can't get things like Plex to work, my NPM reverse proxy goes in/out (external monitoring on my public facing site will cite tons of up/down errors over night when the downloads run) and the only fix is to reboot my server because I can't even just stop the Tube Archivist container. 

My specs are pretty great, so it's not like it's killing the CPU or RAM or anything.
I'm just not sure what's going on, or what I can even do to try and fix it. 
Any advice or suggestions are appreciated! 

I should also note that while the issues were present and I was able to sometimes get the unraid webUI to load, it didn't show any spiked or pegged usages, CPU was at 20% at the time, RAM wasn't full, nothing seemed to be getting used excessively, but pages would fail to load, containers would fail to stop, etc, only recourse was to reboot from command line or web UI if I got lucky.

I tried monitoring things for a bit and it seems to be when it's finalizing/processing the video after it's finished downloading -- the download itself goes fine, and then if I monitor the Downloads page it will hit 100% completion then when it goes to processing the box showing the progress disappears and I lose all visibility on what TA is doing. -- this is also when things start to lock up on the whole server, other dockers become inaccessible, web UI and command line for unraid are terribly slow to respond.

### Relevant log output

```shell
download pending
6a80476e-9d66-4ba5-b673-e55367adc07f: set task id
[2023-01-07 19:00:47,564: INFO/MainProcess] Task download_pending[6a80476e-9d66-4ba5-b673-e55367adc07f] received
[2023-01-07 19:00:47,574: WARNING/ForkPoolWorker-1] Did not acquire download lock.
[2023-01-07 19:00:47,575: INFO/ForkPoolWorker-1] Task download_pending[6a80476e-9d66-4ba5-b673-e55367adc07f] succeeded in 0.006124248000560328s: None
download pending
0a949b28-3747-428b-9b05-729d0d8e24df: set task id
[2023-01-07 19:00:49,529: INFO/MainProcess] Task download_pending[0a949b28-3747-428b-9b05-729d0d8e24df] received
[2023-01-07 19:00:49,534: WARNING/ForkPoolWorker-1] Did not acquire download lock.
[2023-01-07 19:00:49,535: INFO/ForkPoolWorker-1] Task download_pending[0a949b28-3747-428b-9b05-729d0d8e24df] succeeded in 0.0038173840002855286s: None
{"error":{"root_cause":[{"type":"resource_not_found_exception","reason":"snapshot lifecycle policy or policies [ta_daily] not found, no policies are configured"}],"type":"resource_not_found_exception","reason":"snapshot lifecycle policy or policies [ta_daily] not found, no policies are configured"},"status":404}
Filter download view ignored only: False
Filter download view ignored only: True
Filter download view ignored only: True
Filter download view ignored only: False
Filter download view ignored only: True
zB-BRBmv34k: add single vid to download queue
Filter download view ignored only: False
Filter download view ignored only: True
Filter download view ignored only: False
[2023-01-07 19:08:12,452: WARNING/ForkPoolWorker-33] 6q6SXzmcqPY: get metadata from youtube
[2023-01-07 19:08:16,634: WARNING/ForkPoolWorker-33] UCXnNibvR_YIdyPs8PZIBoEw: get metadata from es
[2023-01-07 19:08:17,008: WARNING/ForkPoolWorker-33] 6q6SXzmcqPY: get ryd stats
```

### Anything else?

Host Specs
Processor: Dual Intel® Xeon® CPU E5-2650 v2 @ 2.60GHz
RAM: 128GB RAM
All docker containers run thru a single NIC interface eth1 - no VALNs on it, no bonding, they're just all using that single NIC plain and simple.

**Comment by bbilly1** [comment]
> it will hit 100% completion then when it goes to processing the box showing the progress disappears and I lose all visibility on what TA is doing.

At that step Tube Archivist will move the completed video file from `/cache/download` to the archive folder `/youtube/<channel>/<video>.mp4/` so I'd guess there is an issue there? How did you set this up? Is all on the same host, just between drives? Between what kind of filesystems? Or is there an additional networking layer between? Could also be a faulty drive, raid card...

I recall having a similar difficult to debug issue once, where different virtualization software were trying to use the same network range, so I'd also check that there isn't any overlap there.

But yeah, that is difficult to debug when sitting in front of the machine, but even more so from afar. 

> snapshot lifecycle policy or policies [ta_daily] not found, no policies are configured  

Maybe you activated snapshots without configuring the required env var? [wiki](https://github.com/tubearchivist/tubearchivist/wiki/Settings#snapshots)

**Comment by CorneliousJD** [comment]
> > it will hit 100% completion then when it goes to processing the box showing the progress disappears and I lose all visibility on what TA is doing.
> 
> At that step Tube Archivist will move the completed video file from `/cache/download` to the archive folder `/youtube/<channel>/<video>.mp4/` so I'd guess there is an issue there? How did you set this up? Is all on the same host, just between drives? Between what kind of filesystems? Or is there an additional networking layer between? Could also be a faulty drive, raid card...

So it's all on the same server, just the /cache/download folder is on a RAID1 mirror of SSDs and the final /youtube/ storage path is on the array of unraid, so it is moving from btrfs (cache drive) to xfs (unraid array)
I am currently trying to make the downloads go directly to the array as well so it doesn't have to change filesystems at all and moves from just one folder on the array to another. 

> I recall having a similar difficult to debug issue once, where different virtualization software were trying to use the same network range, so I'd also check that there isn't any overlap there.
> 
> But yeah, that is difficult to debug when sitting in front of the machine, but even more so from afar.

So thankfully no different virtualization software or anything, no different hosts/networks/etc so I should be able to rule this issue out right away. 
 
> > snapshot lifecycle policy or policies [ta_daily] not found, no policies are configured
> 
> Maybe you activated snapshots without configuring the required env var? [wiki](https://github.com/tubearchivist/tubearchivist/wiki/Settings#snapshots)

This is also weird because I have the proper ENV var setup in ES but I did NOT enable snapshots at all.
In fact I just verified they were set to false, and swapped it to "Disable snapshots" and now it's trying to create the initial snapshot? Very weird. It's like this setting is not listening to me properly? But alas, I'll end up having snapshots enabled once I get this other issue sorted out.
![image](https://user-images.githubusercontent.com/724777/211205415-44315c9e-8209-42ad-a389-417750bcf150.png)


Also, thank you so much for taking the time to reply. I'll continue testing the above on my end and let you and any future visitors here if there are any improvements that I see.

If there's any other logs that might indicate what is happening or where things are getting stuck, let me know and I can enable and share them, etc.

**Comment by CorneliousJD** [comment]
Update: Things may actually be fixed now.
I change the /cache/download path so it's on the same array/filesystem as the final /youtube/ storage destination. -- Here's my findings as I tested that.

So once the download itself and the processing box just disappears, it looks like this.
(The box showing progress/processing just disappears) 
This still seems like a bug perhaps that it disappears while processing? 
![image](https://user-images.githubusercontent.com/724777/211205978-12e6e092-0214-4241-a91e-3700c0bd6c9f.png)

And the download folder itself has the following in it.
![image](https://user-images.githubusercontent.com/724777/211205888-8d79a2ef-5a46-4b30-82d3-2068852c2d1f.png)

After it sits there for a moment the .mp4 itself and the .meta file appears and its tarts working on a new .memp.mp4 file 
![image](https://user-images.githubusercontent.com/724777/211205955-0d42f9d6-ac3b-45d3-addc-56c59045e438.png)

**_This is where I think it would lock up/freeze before - somewhere during these steps_**

Which will eventually become a tiny webp, large mp4, and a small png
![image](https://user-images.githubusercontent.com/724777/211206025-f568c2ff-9d5e-4714-9702-e172bae68e05.png)

And then finally it will show just the (what I assume is the final) .mp4 file and the log updates to say "get ryd stats" and the download folder shows this.

![image](https://user-images.githubusercontent.com/724777/211206112-f6086a0a-bfc1-4ba5-95e1-f2958ae5019c.png)

After some time sitting like this, it seems to successfully move the file and move onto the next one now. :) 
I'm letting it chug away and have now gotten through 4 videos without server lockups. Going to let it go through the queue, I'll return here w/ an update once I know if this definitively fixed it or not!

**Comment by CorneliousJD** [comment]
I'm pre-emptively calling this fixed because it's been going for over 2 hours now without locking up -- I certainly never got that yesterday when doing all my testing.

If anyone finds this in the future and has the same issue on unraid, then I added this path to resolve it, which prevents the download files from being put on your btrfs cache drive(s) and having to move them to the array -- this allows it to download straight to the array and then it will just move it to the other path when finished -- still on the same array and filesystem.

Thank you billy1 for the the idea on this.
![image](https://user-images.githubusercontent.com/724777/211212783-a4bfca76-8068-4982-af12-d3df27037ce4.png)

**Comment by bbilly1** [comment]
Excellent write up, thanks for putting in all the details. 

For future readers: The solution above will have some downsides if your `/youtube` is on a different slower network drive, as this will result in additional unnecessary read and writes over a slower connection to first write the video stream, then the audio, then getting it back and combine them and write the final mp4. That's the reasoning why things first go to `/cache/download` and only write the final mp4 to the archive. 

But that obviously doesn't apply if everything is on the same host anyways.

Strange that a simple copy of files can lock up an operating system like that... But I'm glad your solution is working.

**Comment by CorneliousJD** [comment]
Just to add a bit more here -- 100% fixed for sure now. I've downloaded multiple hundred videos over the course of the day with no issues.

I assume I have some unique configuration somewhere going on that's causing the issue with the docker containers and moving of the files between the two filesystems. I have others using unraid that are running TA without any problems and have tens of thousands of videos and did NOT have to make the same change as me, but at least we found the fix rather quickly here and I hope the details help someone else in the future :)
